File: cpp/quantum/src/quantum_gate.cpp
// quantum_gate.cpp

#include "quantum_gate.h" 
#include "qubit_system.h" 
#include "util.h"
#include <cmath>
#include <stdexcept>
using std::complex; 
using std::string;
using std::length_error;
using std::runtime_error;
using std::invalid_argument;

namespace quantum_algorithm_simulator {

// Constructs a new quantum gate operating on an n-qubit system, where the 
// unitary matrix describing the action of the gate is given. If byref is
// false, it assigns the given matrix by reference rather than copying its 
// entries by value. 
// By default, the matrix is the identity matrix and byref is true.
QuantumGate::QuantumGate (int n, complex<double> **matrix, bool byref) {
	if (n <= 0) throw invalid_argument("n must be positive"); 
	this->n = n; 
	if (matrix == NULL) {
		matrix = eyec(pow(2, n), pow(2, n));
		byref = true; // pointless to make a copy of a locally-allocated matrix
	}
	this->matrix = byref ? matrix : copym(matrix, pow(2, n), pow(2, n)); 
} 


// Constructs a new quantum gate operating on an n-qubit system, where the 
// unitary matrix describing the action of the gate is given by a vector of
// values in row-major order.  
QuantumGate::QuantumGate (int n, const complex<double> *vector) {
	if (n <= 0) throw invalid_argument("n must be positive"); 
	this->n = n; 
	this->matrix = vec_to_mat(vector, pow(2, n), pow(2, n));
}

// Copy constructor
QuantumGate::QuantumGate (const QuantumGate &orig) {
	this->n = orig.n; 
	this->matrix = copym(orig.matrix, pow(2, this->n)); 
}

// Copy assignment operator
QuantumGate & QuantumGate::operator= (const QuantumGate &orig) {
	// Do nothing if this == orig
	if (this == &orig) return *this;

	// If the n's are unequal, delete and reallocate matrix
	if (this->n != orig.n) {
		this->n = orig.n; 
		deletem(this->matrix, pow(2, this->n));  
		this->matrix = new_cmat(pow(2, this->n), pow(2, this->n)); 
	}

	// Copy coefficients in place
	copym(orig.matrix, this->matrix, (int) pow(2, this->n)); 

	return *this; 
}


// Destructor
QuantumGate::~QuantumGate () {
	deletem(this->matrix, pow(2, this->n)); 
}


// Returns the number of qubits the gate operates on
int QuantumGate::N () const { 
	return this->n; 
}

// Sets the i, j entry of the matrix to the specified value. 
void QuantumGate::set (int i, int j, complex<double> val) { 
	if (i < 0 || pow(2, this->n) <= i || j < 0 || pow(2, this->n) <= j) { 
		throw invalid_argument("index out of range"); 
	}

	this->matrix[i][j] = val; 
}



// Accesses the i,j entry of the matrix. Can only be used as an accessor - 
// cannot be used to set the value.
complex<double> QuantumGate::operator() (int i, int j) const {
	return this->matrix[i][j]; 
}


// Returns the conjugate transpose of the quantum gate, equivalent to the 
// inverse of the gate (since gate matrices are unitary). 
QuantumGate QuantumGate::H () const {
	QuantumGate gate(this->n, conj_transpose(this->matrix, pow(2, this->n)));
	return gate; 
}


// Overloaded/default version of act, applies to first n qubits where n is the
// number of qubits that q acts on. 
void QuantumGate::act (QubitSystem *q) const {
	this->act(q, 1); 
}


// Performs the action of a quantum gate on a qubit system.  
// The coefficients of the product are the result of the left matrix 
// multiplication of the gate by the system q, starting with the qubit in 
// position INDEX (where 1, not 0, is the first bit). E.g., if the gate G
// operates on 2 qubits, then G.act(q, 2) would operate on the 2nd and 3rd
// qubits of the system, e.g. on the 1 bits of a system in state |01100>. 
// 
// It operates IN PLACE on the argument q, so the qubit system is changed
// by operation of the gate. This setup, as opposed to returning a new system as
// the product, prevents violations of the no-cloning theorem. 
void QuantumGate::act (QubitSystem *q, int index) const {
	// Throw error if the dimensions don't fit
	if (index <= 0 || index + this->n - 1 > q->N()) { 
		throw invalid_argument("Dimension mismatch with gate and qubit system"); 
	}
	// Throw error if matrix is not unitary
	if (!is_unitary(this->matrix, pow(2, this->n))) {
		throw runtime_error("Cannot perform non-unitary operation on a qubit");
	}

	complex<double> **mat, **tmp, **I; 

	if (index == 1 && this->n == q->N()) { 
		prod_ip(this->matrix, q->coeffs, pow(2, this->n)); 
	}
	else if (index > 1 && index + this->n - 1 == q->N()) { 
		I = eyec(pow(2, index - 1), pow(2, index - 1)); 
		mat = kronecker(I, this->matrix, pow(2, index - 1), pow(2, index - 1), 
		                pow(2, this->n), pow(2, this->n)); 
		prod_ip(mat, q->coeffs, pow(2, q->N())); 

		deletem(mat, pow(2, q->N())); 
		deletem(I, pow(2, index - 1)); 
	}
	else if (index == 1 && this->n < q->N()) { 
		I = eyec(pow(2, q->N() - this->n), pow(2, q->N() - this->n)); 
		mat = kronecker(this->matrix, I, pow(2, this->n), pow(2, this->n), 
		                pow(2, q->N() - this->n), pow(2, q->N() - this->n)); 
		prod_ip(mat, q->coeffs, pow(2, q->N())); 

		deletem(mat, pow(2, q->N())); 
		deletem(I, pow(2, q->N() - this->n)); 
	}
	else { // if (index > 1 && index + this->n - 1 < q->N())  
		I = eyec(pow(2, index - 1), pow(2, index - 1)); 
		tmp = kronecker(I, this->matrix, pow(2, index - 1), pow(2, index - 1), 
		                pow(2, this->n), pow(2, this->n)); 

		deletem(I, pow(2, index - 1)); 

		I = eyec(pow(2, q->N() - (index + this->n - 1)), 
		         pow(2, q->N() - (index + this->n - 1)));
		mat = kronecker(tmp, I, pow(2, this->n + index - 1),
		                pow(2, this->n + index - 1),
		                pow(2, q->N() - (index + this->n - 1)), 
		                pow(2, q->N() - (index + this->n - 1))); 
		prod_ip(mat, q->coeffs, pow(2, q->N())); 

		deletem(mat, pow(2, q->N())); 
		deletem(tmp, pow(2, this->n + index - 1)); 
		deletem(I, pow(2, q->N() - (index + this->n - 1))); 
	}
}


// Sum of two quantum gates - just the matrix sum of the two matrices. 
QuantumGate & QuantumGate::operator+= (const QuantumGate &other) {
	// Throw error if the lengths differ
	if (this->n != other.n) {
		throw length_error("Cannot add quantum gates of unequal dimension");
	}

	sum_ip(this->matrix, other.matrix, pow(2, this->n)); 
	return *this; 
}


// Difference of two quantum gates - just the difference of the two matrices. 
QuantumGate & QuantumGate::operator-= (const QuantumGate &other) {
	// Throw error if the lengths differ
	if (this->n != other.n) {
		throw length_error("Cannot subtract quantum gates of unequal dimension");
	}

	diff_ip(this->matrix, other.matrix, pow(2, this->n)); 
	return *this; 
}

// Multiplies two gates via standard matrix multiplication. The action of the
// returned gate is equivalent to the action of the righthand gate followed by
// that of the lefthand gate.
QuantumGate & QuantumGate::operator*= (const QuantumGate &other) {
	// Throw error if the lengths differ
	if (this->n != other.n) { 
		throw length_error("Cannot multiply quantum gates of unequal dimension");
	}

	prod_ip(this->matrix, other.matrix, pow(2, this->n));
	return *this; 
}


// Returns the gate that results from taking the matrix product of the given
// gate with itself e times, that is, raising the matrix to the exponent 
QuantumGate & QuantumGate::operator^= (int e) {
	if (e < 0) throw invalid_argument("exponent must be nonnegative"); 
	
	mpow_ip(this->matrix, e, pow(2, this->n)); 
	return *this; 
}

// Sum of two quantum gates - just the matrix sum of the two matrices. 
QuantumGate operator+ (const QuantumGate &g1, const QuantumGate &g2) {
	// Throw error if the lengths differ
	if (g1.n != g2.n) {
		throw length_error("Cannot add quantum gates of unequal dimension");
	}

	QuantumGate g_out(g1.n, sum(g1.matrix, g2.matrix, pow(2, g1.n))); 
	return g_out;
}

// Difference of two quantum gates - just the difference of the two matrices. 
QuantumGate operator- (const QuantumGate &g1, const QuantumGate &g2) {
	// Throw error if the lengths differ
	if (g1.n != g2.n) {
		throw length_error("Cannot subtract quantum gates of unequal dimension");
	}

	QuantumGate g_out(g1.n, diff(g1.matrix, g2.matrix, pow(2, g1.n))); 
	return g_out;
}

// Multiplies two gates via standard matrix multiplication. The action of the
// returned gate is equivalent to the action of the righthand gate followed by
// that of the lefthand gate.
QuantumGate operator* (const QuantumGate &g1, const QuantumGate &g2) {
	// Throw error if the lengths differ
	if (g1.n != g2.n) { 
		throw length_error("Cannot multiply quantum gates of unequal dimension");
	}

	QuantumGate product(g1.n, prod(g1.matrix, g2.matrix, pow(2, g1.n)));
	return product;
}

// Returns the gate that results from taking the matrix product of the given
// gate with itself e times, that is, raising the matrix to the exponent 
QuantumGate operator^ (const QuantumGate &gate, int e) {
	if (e < 0) throw invalid_argument("exponent must be nonnegative"); 
	
	QuantumGate g_out(gate.n, mpow(gate.matrix, e, pow(2, gate.n))); 
	return g_out; 
}

	
// Returns the tensor product, i.e. the Kronecker product, of two quantum gates.
QuantumGate kron (const QuantumGate &g1, const QuantumGate &g2) {
	QuantumGate product(g1.n + g2.n, kronecker(g1.matrix, g2.matrix, pow(2, g1.n),
	                                   pow(2, g1.n), pow(2, g2.n), pow(2, g2.n)));
	return product; 
}


// Returns the gate that results from taking the tensor product of the given
// gate with itself e times, analogous to raising g to an exponent e.  
QuantumGate tensor_pow (const QuantumGate &g, int e) {
	if (e <= 0) throw invalid_argument("exponent must be positive"); 

	const int pow2n = pow(2, g.n); 
	int nprod = pow2n; // dimension of matrix "product"
	complex<double> **product = copym(g.matrix, nprod); 
	complex<double> **tmp;

	for (int i = 2; i <= e; i++) { 
		tmp = copym(product, nprod); 
		deletem(product, nprod); 
		product = kronecker(tmp, g.matrix, nprod, nprod, pow2n, pow2n); 
		deletem(tmp, nprod); 
		nprod *= pow2n; 
	}

	QuantumGate g_out(g.n * e, product); 
	return g_out; 
} 

} 

File: python/quantum/quantum_gate.py
# quantum_gate.py

"""Contains the QuantumGate class and several closely related functions, for
representing and working with quantum logic gates."""

from math import log
import numpy
from numpy import array, copy, eye, linalg

from quantum.qubit_system import QubitSystem
from quantum._util import EPS, is_unitary

class QuantumGate(object):
    """Represents a quantum logic gate.

    The class is a thin wrapper around a Numpy array that represents the action
    of the gate.

    Attributes:
        __matrix: The unitary matrix representing the action of the gate on a
        system of qubits. A quantum gate that operates on n qubits has a matrix
        of dimension 2^n x 2^n.
    """

    def __init__(self, matrix):
        """Constructor. Creates a quantum gate with the given matrix.

        Args:
            matrix: A 2^n x 2^n array-like object, where n is the number of
            quantum bits the gate should operate on.

        Raises:
            ValueError if "matrix" has invalid dimensions for an quantum gate
            matrix.
        """
        # test that matrix is square and with dimensions that are a power of 2
        if matrix.ndim != 2 or matrix.shape[0] != matrix.shape[1] \
           or abs(log(matrix.shape[0], 2) % 1) > EPS:
            raise ValueError('invalid matrix dimensions')
        self.__matrix = array(matrix)


    def __repr__(self):
        return 'QuantumGate(' + repr(self.__matrix) + ')'


    def __str__(self):
        return 'QuantumGate(' + str(self.__matrix) + ')'


    def copy(self):
        """Deep copy"""
        return QuantumGate(self.__matrix)


    def n(self):
        """Number of qubits that the quantum gate acts on."""
        return int(round(log(self.__matrix.shape[0], 2)))


    # Functions for accessing and setting the matrix
    def matrix(self):
        """Returns the matrix that defines the action of the quantum gate."""
        return self.__matrix

    def __getitem__(self, key):
        return self.__matrix[key]

    def __setitem__(self, key, value):
        self.__matrix[key] = value


    def act(self, qubits, index=1):
        """Performs the action of a quantum gate on a qubit system.

        Operates in-place on the system "qubits", so the original system is
        changed by interaction with the gate. This avoids violations of the no-
        cloning theorem.

        Args:
            qubits: A system of qubits for the gate to act on.
            index: Starting index of the first qubit in the system for the gate
            to act one. E.g. if the gate acts on two qubits and index = 3, then
            the gate would act on the 3rd and 4th qubits in the system (where
            qubits are indexed starting at one).

        Raises:
            ValueError if there is a dimension mismatch between the gate and the
            qubits to be operated on.
            RuntimeError if the matrix is not unitary.
        """
        if index <= 0 or index + self.n() - 1 > qubits.n():
            raise ValueError('Dimension mismatch with gate and qubit system')
        if not is_unitary(self.__matrix):
            raise RuntimeError('Non-unitary matrix')

        # construct a matrix to operate only on the desired qubits
        G = copy(self.__matrix)
        if index > 1:
            G = numpy.kron(eye(2**(index - 1)), G)
        if index + self.n() - 1 < qubits.n():
            G = numpy.kron(G, eye(2**(qubits.n() - (index + self.n() -1))))

        qubits._QubitSystem__coeffs = numpy.dot(G, qubits._QubitSystem__coeffs)


    def H(self):
        """Returns a new quantum gate whose matrix is the conjugate transpose of
        the current gate's matrix."""
        return QuantumGate(self.__matrix.conj().T)

    def abs(self):
        """Returns a new quantum gate whose matrix is the (entrywise) absolute
        value of the current gate's matrix."""
        return QuantumGate(abs(self.__matrix))

    def __pos__(self):
        """Returns a copy of the object unchanged."""
        return copy(self)

    def __neg__(self):
        """Negates the gate's matrix entrywise."""
        return QuantumGate(-self.__matrix)

    def __add__(self, other):
        """Returns a quantum gate created by adding the matrices of two gates,
        via ordinary matrix addition."""
        return QuantumGate(self.__matrix + other.__matrix)

    def __sub__(self, other):
        """Returns a quantum gate created by subtracting the matrices of two
        gates, via ordinary matrix subtraction."""
        return QuantumGate(self.__matrix - other.__matrix)

    # Entrywise matrix multiplication - also shorthand for self.act(other) where
    # other is a qubit system.
    def __mul__(self, other):
        """Returns a quantum gate created by multiplying the matrices of two
        gates, via entrywise multiplication."""
        if isinstance(other, QubitSystem):
            return self.act(other)
        else:
            return QuantumGate(self.__matrix - other.__matrix)

    def __pow__(self, e):
        """Returns a quantum gate created by raising the matrix of the gate to a
        power, using entrywise exponentiation."""
        return QuantumGate(self.__matrix**e)

    # Comparison operators
    def __lt__(self, other):
        return self.__matrix < other.__matrix
    def __le__(self, other):
        return self.__matrix <= other.__matrix
    def __eq__(self, other):
        return self.__matrix == other.__matrix
    def __ne__(self, other):
        return self.__matrix != other.__matrix
    def __gt__(self, other):
        return self.__matrix > other.__matrix
    def __ge__(self, other):
        return self.__matrix >= other.__matrix



def dot(A, B):
    """Returns a quantum gate created by multiplying the matrices of two gates,
    via matrix multiplication."""
    return QuantumGate(numpy.dot(A.matrix(), B.matrix()))

def matrix_power(G, e):
    """Returns a quantum gate created by raising the matrix of the gate to an
    integer power, using matrix exponentiation."""
    return QuantumGate(linalg.matrix_power(G.matrix(), e))

def kron(A, B):
    """Returns a quantum gate created by taking the Kronecker product (tensor
    product) of two other gates."""
    return QuantumGate(numpy.kron(A.matrix(), B.matrix()))

def tensor_power(G, e):
    """Returns the gate that results from taking the tensor product of the given
    gate with itself e times, analogous to raising the gate to an exponent e."""
    if e < 1 or e % 1 != 0:
        raise ValueError('Exponent must be a nonnegative integer')

    M = copy(G.matrix())
    for _ in range(1, e):
        M = numpy.kron(M, G.matrix())
    return QuantumGate(M)





